packages/babel-parser/src/tokenizer/state.js

```diff
@@ -178,3 +178,14 @@ export default class State {
     return state;
   }
 }
+
+export type LookaheadState = {
+  pos: number,
+  value: any,
+  type: TokenType,
+  start: number,
+  end: number,
+  /* Used only in readSlashToken */
+  exprAllowed: boolean,
```

Comment:
Can be 5% faster if we get rid of `exprAllowed` and `inType`, so that they don't have to be copied.

Reply:
Does copying a couple of bool influence that much perf?

Reply:
Yeah. Unlike C, in V8 the boolean literal occupies 64 bit memory, essentially an address points to the `true`/`false` built-ins. And a sequence of boolean literals will not be compressed into a bit array. So two boolean literal = 16 byte memory.

In Babel 7.7 we improved the traverser performance by compressing 3 booleans to a bit array: https://hackmd.io/UMdqwvVgQGaofjCZHfGKrA?view#Compress-boolean-flags

Note that the benchmark is specifically constructed to highlight the performance impact of `lookahead()`, so in the real word they are not that significant. However, the performance improvement is a long-term task and eventually every tiny improvement counts.

Reply:
@JLHwung Funny to see devs playing with perf :) Your suggested changed is only micro-optimization and not scalable, but yeah you are on right track.
The entire tokenizer is an `Class` that extends `ParserErrors `.  Run a benchmark and check the `extends` perf and you get chocked about perf loss.
Each token is it's own `Class`. Try flatten it instead into an infinite state machine. I'm sure you gain 20% perf. As it's now it consumes memory.

You really should reduce amount of object access too. This is expensive
` this.state.lastTokEnd`  vs  `this.lastTokEnd`. Why do you need this extra `state` object inside the class?
A quick google search will also let you know that `this` is expensive. 

There are many other things you can do to optimize Babel parser. I mentioned only a few.